test/util.sh

```diff
@@ -0,0 +1,203 @@
+}
+
+# expect_failure_and_text runs the cmd and expects a non-zero exit code
+# as well as running a grep test to find the given string in the output
+function os::test::expect_failure_and_text() {
+	cmd=$1
+	expected_text=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "failure_func" "${expected_text}"
+}
+
+# expect_success_and_not_text runs the cmd and expects an exit code of 0
+# as well as running a grep test to ensure the given string is not in the output
+function os::test::expect_success_and_not_text() {
+	cmd=$1
+	expected_text=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "success_func" "${expected_text}" "failure_func"
+}
+
+# expect_failure_and_not_text runs the cmd and expects a non-zero exit code
+# as well as running a grep test to ensure the given string is not in the output
+function os::test::expect_failure_and_not_text() {
+	cmd=$1
+	expected_text=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "failure_func" "${expected_text}" "failure_func"
+}
+
+# MEDIUM-LEVEL WRAPPER FUNCTIONS: ASSUME NO CODES
+
+# expect_code runs the cmd and expects a given exit code
+function os::test::expect_code() {
+	cmd=$1
+	expected_cmd_code=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "specific_code_func ${expected_cmd_code}"
+}
+
+# expect_code_and_text runs the cmd and expects the given exit code
+# as well as running a grep test to find the given string in the output
+function os::test::expect_code_and_text() {
+	cmd=$1
+	expected_cmd_code=$2
+	expected_text=$3
+
+	os::test::expect_exit_code_run_grep "${cmd}" "specific_code_func ${expected_cmd_code}" "${expected_text}"
+}
+
+# expect_code_and_not_text runs the cmd and expects the given exit code
+# as well as running a grep test to ensure the given string is not in the output
+function os::test::expect_code_and_not_text() {
+	cmd=$1
+	expected_cmd_code=$2
+	expected_text=$3
+
+	os::test::expect_exit_code_run_grep "${cmd}" "specific_code_func ${expected_cmd_code}" "${expected_text}" "failure_func"
+}
+
+# TEST EXECUTION FUNCTION
+
+# expect_exit_code_and_text runs the provided test command and expects a specific exit code from that
+# command as well as the success of a specified `grep` invocation. Output from the command to be 
+# tested is suppressed unless either `VERBOSE=1` or the test fails.
+function os::test::expect_exit_code_run_grep() {
+	cmd=$1
+	# default expected cmd code to 0 for success
+	cmd_eval_func=${2:-success_func}
+	# default grep args to a something that always passes, even for zero-length input
+	grep_args=${3:-'$'} 
+	# default expected test code to 0 for success
+	test_eval_func=${4:-success_func}
+
+	echo "Running  ${cmd}..."
+	
+	set +o errexit
+	cmd_output=$(${cmd} 2>&1)
+	cmd_result=$?
+	cmd_succeeded=$( ${cmd_eval_func} "${cmd_result}"; echo $? )
+
+	test_output=$(echo "${cmd_output}" | grep -E "${grep_args}" 2>&1)
+	test_result=$?
+	test_succeeded=$( ${test_eval_func} "${test_result}"; echo $? )
+	set -o errexit
```

Comment:
This can't be done here.  I don't remember seeing it before. There's no guarantee that the caller hadn't already called `set +e` in their script.